src/pages/Storefront/Cart.js

```diff
@@ -14,23 +17,102 @@ import {
-import CartSkeleton from 'components/storefront/Cart/Layout/CartSkeleton';
+import { ImPriceTags } from 'react-icons/im';
+
+const calculateDiscount = (discountApplications, lineItems) => {
+  let discountVal = 0;
+  let shippingCost = 'TBD';
+  let discountType = '';
+  if (discountApplications.length > 0) {
+    const appliedDiscount = discountApplications[0];
+    discountType = appliedDiscount.targetType;
+    if (discountType === 'SHIPPING_LINE') {
+      shippingCost = 'FREE';
+    } else if (discountType === 'LINE_ITEM') {
+      discountVal = lineItems
+        .reduce((acc, lineItem) => {
+          return lineItem.discountAllocations
+            ? acc +
+                lineItem.discountAllocations.reduce((acc2, discAlloc) => {
+                  return acc2 + parseFloat(discAlloc.allocatedAmount.amount);
+                }, 0)
+            : acc;
+        }, 0)
+        .toFixed(2);
+    }
+  }
+  return { discountVal, shippingCost, discountType };
+};
+
+const DiscountMsg = ({ discountApplications, discountSuccess, removeCoupon }) => {
+  const msg =
+    discountSuccess === 'SUCCESS' ? (
+      <chakra.h4 whiteSpace="nowrap" textStyle="lightCaption">
+        Success! Your code <b>{discountApplications[0].code}</b> has been applied.
+      </chakra.h4>
+    ) : discountSuccess === 'INVALID' ? (
+      <chakra.h4 color="brand.red" textStyle="lightCaption">
+        Sorry, that coupon code is not applicable.
+      </chakra.h4>
+    ) : null;
+
+  return (
+    <Box pt="4">
+      {msg}
+      {discountApplications && discountApplications[0] && (
+        <Tag size="lg" mt="2">
+          <TagLeftIcon boxSize="12px" as={ImPriceTags} />
+          <TagLabel>{discountApplications[0].code}</TagLabel>
+          <TagCloseButton onClick={removeCoupon} />
+        </Tag>
+      )}
+    </Box>
+  );
+};
 
 const CartItems = ({ checkoutData }) => {
-  const { id: checkoutId, lineItems } = checkoutData;
+  const { id: checkoutId, lineItems, discountApplications } = checkoutData;
+  const { addDiscount, removeDiscount } = useShopify();
   const cartItemsCount = lineItems.reduce((acc, cur) => acc + cur.quantity, 0);
+  const [discountCode, setDiscountCode] = useState('');
+  // discountSuccess is enum "INITIAL" | "INVALID" | "SUCCESS"
+  const [discountSuccess, setDiscountSuccess] = useState('INITIAL');
+
+  const onChangeCoupon = e => {
+    // Assuming that all discounts are strictly uppercase
+    setDiscountCode(e.target.value.toUpperCase());
+  };
+
+  const applyCoupon = async e => {
+    e.preventDefault();
+    const res = await addDiscount(checkoutId, discountCode);
+
+    // TO DO: Proper error handling (catch error in reducer?)
+    if (res.discountApplications.length > 0) {
+      // Handling case where there is an existing coupon + user enters invalid coupon
+      if (discountCode !== res.discountApplications[0].code) {
+        setDiscountSuccess('INVALID');
```

Comment:
I removed the call to `removeCoupon` here, choosing to preserve the previous coupon instead. This is what Shopify does in the checkout screen as well